syntax = "proto3";

package mantenedor_productos;

service ProductMaintainer {
  rpc GetProduct(GetProductRequest) returns (Product);
  rpc CreateProduct(CreateProductRequest) returns (Product);
  rpc UpdateProduct(UpdateProductRequest) returns (Product);
  rpc DeleteProduct(DeleteProductRequest) returns (DeleteProductResponse);
  rpc UploadProductImage(stream UploadProductImageRequest) returns (UploadProductImageResponse);
  rpc GetAllProducts(GetAllProductsRequest) returns (ProductList);
}

message Product {
  int32 id = 1;
  string nombre = 2;
  float precio = 3;
  string imagen = 4; // URL o la ruta de la imagen
}

message GetProductRequest {
  int32 product_id = 1;
}

message CreateProductRequest {
  string nombre = 1;
  float precio = 2;
  string imagen = 3;
}

message UpdateProductRequest {
  int32 id = 1;
  string nombre = 2;
  float precio = 3;
  string imagen = 4;
}

message DeleteProductRequest {
  int32 product_id = 1;
}

message DeleteProductResponse {
  bool success = 1;
  string message = 2;
}

message UploadProductImageRequest {
  int32 product_id = 1;
  bytes chunk_data = 2; // Datos binarios de la imagen
  string filename = 3; // Nombre de archivo original para la extensi√≥n
}

message UploadProductImageResponse {
  bool success = 1;
  string image_url = 2; 
  string message = 3;
}

message GetAllProductsRequest {
  
}

message ProductList {
  repeated Product products = 1;
}